src/NUnitFramework/framework/Internal/Filters/InFilter.cs

```diff
@@ -0,0 +1,100 @@
+            _selector = selector;
+            _values = new HashSet<string?>(values);
+        }
+
+        /// <summary>
+        /// Tries to optimize OrFilter into InFilter if contained filters are all same and non-regex.
+        /// </summary>
+        public static bool TryOptimize(OrFilter orFilter, [NotNullWhen(true)] out InFilter? optimized)
+        {
+            // check if we have all filters accessing same field in OR fashion without regex
+            var hashCommonTypeAndNonRegex = orFilter.Filters.Count > 0;
+            Type? commonType = null;
+
+            // eagerly build expected value collection
+            var expectedValues = new List<string?>(orFilter.Filters.Count);
+
+            // use for for faster traversal without boxed enumerator
+            for (var i = 0; i < orFilter.Filters.Count; i++)
+            {
+                var filter = orFilter.Filters[i];
+
+                // make sure invariants are valid
+                commonType ??= filter.GetType();
+                hashCommonTypeAndNonRegex &= commonType == filter.GetType();
+
+                // our supported filter types
+                hashCommonTypeAndNonRegex &=
+                    filter is ClassNameFilter { IsRegex: false }
+                    || filter is FullNameFilter { IsRegex: false }
+                    || filter is IdFilter { IsRegex: false }
+                    || filter is MethodNameFilter { IsRegex: false }
+                    || filter is NamespaceFilter { IsRegex: false }
+                    || filter is TestNameFilter { IsRegex: false };
+
+                if (!hashCommonTypeAndNonRegex)
+                {
+                    break;
+                }
+
+                expectedValues.Add(((ValueMatchFilter) filter).ExpectedValue);
+            }
+
+            if (hashCommonTypeAndNonRegex)
+            {
+                // we can transform, all filters in non-empty collection are of same type
+                Func<ITest, string?> selector = orFilter.Filters[0] switch
+                {
+                    ClassNameFilter _ => test => test.ClassName,
+                    FullNameFilter _ => test => test.FullName,
+                    IdFilter _ => test => test.Id,
+                    MethodNameFilter _ => test => test.MethodName,
+                    NamespaceFilter _ => test => test.TypeInfo?.Namespace,
+                    TestNameFilter _ => test => test.Name,
+                    _ => throw new InvalidOperationException("Invalid filter, logic wrong")
+                };
+                optimized = new InFilter(selector, expectedValues);
+                return true;
+            }
+
+            optimized = default;
+            return false;
+        }
+
+        public override bool Match(ITest test)
+        {
+            return _values.Contains(_selector(test));
+        }
+
+        public override TNode AddToXml(TNode parentNode, bool recursive)
+        {
+            throw new NotImplementedException();
```

Comment:
I think this will give us problems in runners that also write the filters in the XML (e.g. NUnitLite, the console etc). I guess we could write the filter as a `OrFilter` (https://docs.nunit.org/articles/nunit/technical-notes/usage/Test-Filters.html), but then we need to store the original OrFilter also (or at least have more information).

Reply:
My thinking was that this is dynamic, e.g. translated runtime so it would not matter. I probably don't understand the full ramifications, should this be addressed in this PR, will this be breaking things?